src/node_constants.cc

```diff
@@ -795,11 +795,39 @@ void DefinePriorityConstants(Local<Object> target) {
 #endif
 }
 
+// Convert nid's to the string representation of their OID. Non-reentrant, and
+// will abort if called with invalid nids (so only pass values from OpenSSL's
+// headers).
+static const char* OBJ_nid2oid(int nid) {
+  ASN1_OBJECT* obj = OBJ_nid2obj(nid);
+  CHECK(obj);
+
+  static char buf[128];
+  CHECK_LE(OBJ_obj2txt(buf, sizeof(buf), obj, 1), 128);
+
+  return buf;
+}
+
 void DefineCryptoConstants(Local<Object> target) {
 #ifdef OPENSSL_VERSION_NUMBER
     NODE_DEFINE_CONSTANT(target, OPENSSL_VERSION_NUMBER);
 #endif
 
+#define NID2OID(nid) NODE_DEFINE_STRING_CONSTANT(target, #nid, OBJ_nid2oid(nid))
+
+  NID2OID(EVP_PKEY_RSA);
+  NID2OID(EVP_PKEY_RSA_PSS);
+  NID2OID(EVP_PKEY_DSA);
+  NID2OID(EVP_PKEY_DH);
+  NID2OID(EVP_PKEY_EC);
+  // Note for backporters: following are new in openssl 1.1.1.
+  NID2OID(EVP_PKEY_ED25519);
```

Comment:
Wrap these in `#ifdef EVP_PKEY_ED25519` etc.?

Reply:
master doesn't support anything earlier than openssl 1.1.1, and I didn't want to introduce the possibility that someone might try and get a superficially working build. This is constructed this way (without the `#define EVPS(V) V( EVP_PKEY_DH)` idiom so that when this backports to 11.x or 10.x, it will be trivial to put in the macros.